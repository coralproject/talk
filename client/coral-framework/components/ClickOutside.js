import { Component, cloneElement, Children } from 'react';
import PropTypes from 'prop-types';
import { findDOMNode } from 'react-dom';

export default class ClickOutside extends Component {
  static propTypes = {
    onClickOutside: PropTypes.func,
    children: PropTypes.node,
  };

  static contextTypes = {
    pym: PropTypes.object,
  };

  domNode = null;

  handleClick = e => {
    const { onClickOutside } = this.props;
    if (!e || !this.domNode.contains(e.target)) {
      onClickOutside && onClickOutside(e);
    }
  };

  componentDidMount() {
    const { pym } = this.context;
    this.domNode = findDOMNode(this);
    document.addEventListener('click', this.handleClick, true);
    if (pym) {
      pym.onMessage('click', this.handleClick);
    }
  }

  componentWillUnmount() {
    const { pym } = this.context;
    document.removeEventListener('click', this.handleClick, true);
    if (pym) {
      pym.messageHandlers.click = pym.messageHandlers.click.filter(
        h => h !== this.handleClick
      );
    }
  }

  render() {
    const { children, onClickOutside: _, ...rest } = this.props;
    return cloneElement(Children.only(children), rest);
  }
}
