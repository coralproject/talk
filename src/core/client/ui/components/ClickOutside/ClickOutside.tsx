import React from "react";
import { findDOMNode } from "react-dom";

interface Props {
  onClickOutside: () => void;
  children: React.ReactNode;
}

class ClickOutside extends React.Component<Props> {
  public domNode: Element | null = null;

  public handleClick = (e: MouseEvent) => {
    const { onClickOutside } = this.props;
    if (!e || !this.domNode!.contains(e.target as HTMLInputElement)) {
      // tslint:disable-next-line:no-unused-expression
      onClickOutside && onClickOutside();
    }
  };

  public componentDidMount() {
    this.domNode = findDOMNode(this) as Element;
    document.addEventListener("click", this.handleClick, true);
  }

  public componentWillUnmount() {
    document.removeEventListener("click", this.handleClick, true);
  }

  public render() {
    return this.props.children;
  }
}
export default ClickOutside;
