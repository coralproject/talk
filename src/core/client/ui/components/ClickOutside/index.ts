export { default as ClickOutside, ClickFarAwayRegister } from "./ClickOutside";
