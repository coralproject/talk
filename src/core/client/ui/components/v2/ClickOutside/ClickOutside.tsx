import React, { FunctionComponent } from "react";

import UIContext from "../UIContext";

export interface ClickOutsideProps {
  onClickOutside: (e?: MouseEvent) => void;

  /**
   * Allow you to change the `Window` reference.
   */
  window?: Window;

  children: React.ReactNode;
}

export class ClickOutside extends React.Component<ClickOutsideProps> {
  public static defaultProps: Partial<ClickOutsideProps> = {
    // eslint-disable-next-line no-restricted-globals
    window,
  };

  public domNode: Element | null = null;

  public handleRef = (e: Element | null) => {
    this.domNode = e;
  };

  public handleClick = (e: MouseEvent) => {
    const path =
      (e.composedPath && e.composedPath()) ||
      // Supports older browsers.
      (e as any).path;
    const { onClickOutside } = this.props;
    if (this.domNode && !this.domNode.contains(path[0] as HTMLInputElement)) {
      // eslint-disable-next-line no-unused-expressions
      onClickOutside && onClickOutside(e);
    }
  };

  public componentDidMount() {
    this.props.window!.document.addEventListener(
      "click",
      this.handleClick,
      true
    );
  }

  public componentWillUnmount() {
    this.props.window!.document.removeEventListener(
      "click",
      this.handleClick,
      true
    );
  }

  public render() {
    return (
      <div ref={(node) => (this.domNode = node)}>{this.props.children}</div>
    );
  }
}

const ClickOutsideWithContext: FunctionComponent<ClickOutsideProps> = (
  props
) => (
  <UIContext.Consumer>
    {({ renderWindow }) => <ClickOutside {...props} window={renderWindow} />}
  </UIContext.Consumer>
);

export default ClickOutsideWithContext;
