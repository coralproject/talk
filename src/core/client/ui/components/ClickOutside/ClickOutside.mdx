---
name: ClickOutside
menu: UI Kit
---

import { Playground, PropsTable } from 'docz'
import ClickOutside from './ClickOutside'
import Button from '../Button/Button'

# ClickOutside

A Component to handle click events outside the children components.

## Basic usage
Wrap a Component with `<ClickOutside>` and pass a function to `onClickOutside`. This function will trigger when the user
clicks outside the component.

### Test
Click the blue background. It should trigger an alert. Nothing should happen if you click the button.

<Playground>
  <div style={{background: 'blue', padding: '10px'}}>
    <ClickOutside onClickOutside={() => alert('You clicked outside!')}>
      <Button variant="filled">Push Me</Button>
    </ClickOutside>
  </div>
</Playground>
