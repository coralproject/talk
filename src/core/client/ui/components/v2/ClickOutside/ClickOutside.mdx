---
name: ClickOutside
menu: UI Kit
---

import { Playground, PropsTable } from 'docz'
import ClickOutside from './ClickOutside'
import Button from '../Button/Button'

# ClickOutside

A Component to handle click events outside the children components.

## Basic usage
Wrap a Component with `<ClickOutside>` and pass a function to `onClickOutside`. This function will trigger when the user
clicks outside the component.

### Test
Click the blue background. It should trigger an alert. Nothing should happen if you click the button.

<Playground>
  <div id="outside" style={{background: 'blue', padding: '10px'}}>
    <ClickOutside onClickOutside={e => {
      if (e.srcElement.id === "outside") {
        alert('You clicked outside!');
      }}}>
      <Button variant="filled">Push Me</Button>
    </ClickOutside>
  </div>
</Playground>
