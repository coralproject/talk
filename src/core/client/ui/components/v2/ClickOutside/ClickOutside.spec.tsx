import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import sinon from "sinon";

import { ClickOutside } from "./ClickOutside";

it("should render correctly", () => {
  const noop = () => null;
  render(
    <ClickOutside onClickOutside={noop}>
      <span>Hello World!</span>
    </ClickOutside>
  );
  expect(screen.getByText("Hello World!"));
});

it("should detect click outside", () => {
  const onClickOutside = sinon.spy();
  render(
    <div data-testid="outside">
      <ClickOutside onClickOutside={onClickOutside}>
        <span>Hello World!</span>
      </ClickOutside>
    </div>
  );
  const outsideDiv = screen.getByTestId("outside");
  userEvent.click(outsideDiv);

  expect(onClickOutside.calledOnce).toEqual(true);
});

it("should ignore click inside", () => {
  const onClickOutside = sinon.spy();
  render(
    <ClickOutside onClickOutside={onClickOutside}>
      <button data-testid="click-outside-test-button">Push Me</button>
    </ClickOutside>
  );

  const insideButton = screen.getByTestId("click-outside-test-button");
  userEvent.click(insideButton);
  expect(onClickOutside.calledOnce).toEqual(false);
});
