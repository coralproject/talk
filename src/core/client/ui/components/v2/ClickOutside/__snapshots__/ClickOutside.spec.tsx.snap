// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `"<span>Hello World!</span>"`;
